File: Day_50/Subnetting.py
def ipv4_mask(cidr):
    mask = 2**32 - 2**(32- int(cidr))
    return (mask >>sh & 0xff for sh in (24,16,8,0))

def ipv4_bytes(ip):
    return (int(b) for b in ip.split('.'))

def ipv4_cidr_member_of(ip1, ip2):
    ip2, m = ip2.split('/')
    return not any((a ^ b) & m 
                   for a, b, m in 
                   zip(ipv4_bytes(ip1), 
                       ipv4_bytes(ip2), 
                       ipv4_mask(m)))

def ipv4_cidr_range_bytes(ip):
    ip, m = ip.split('/')
    ip    = list(ipv4_bytes(ip))
    m     = list(ipv4_mask(m))
    start = [ b &  m         for b, m in zip(ip, m)]
    end   = [(b | ~m) & 0xff for b, m in zip(ip, m)]
    return start, end

#print(ipv4_mask("2.225.255.0"))

tempString = input()
n,k = tempString.split(" ")
n = int(n)
k= int(k)
netMask = list()
ipEnd = list()
location = list()

for i in range(n):
    inputString = input()
    ipl = inputString.split(" ")
    netMask.append(ipl[0].split("/")[0])
    ipEnd.append(ipl[1].split("/")[0])
    location.append(ipl[2])


for i in range(k):
    ipToTest = input()
    flag = 0
    for i in range(n):
        if(ipv4_cidr_member_of(ipToTest,netMask[i])):
            print(location[i])
            flag =1
            break
    if flag == 1:
        print("None")
       
        

        

File: Day_50/Subnetting.java
import java.net.*; 
import java.io.*; 
import java.util.*; 
import java.net.InetAddress; 
public class Subnetting {
    public static long ipToLong(InetAddress ip) {
        byte[] octets = ip.getAddress();
        long result = 0;
        for (byte octet : octets) {
            result <<= 8;
            result |= octet & 0xff;
        }
        return result;
    }

    public static void main(String[] args) throws UnknownHostException {
        Scanner Input = new Scanner(System.in);
        int t;
        int q;

        t=Input.nextInt();
        q=Input.nextInt();

        ArrayList<String>ipl = new ArrayList<String>();
        ArrayList<String>iph = new ArrayList<String>();
        ArrayList<String>loc = new ArrayList<String>();

        Input.nextLine();
        for(int i=0;i<t;i++){
            String l = Input.nextLine();
            String[] s = l.split(" ");
            String location = s[2];
            String ip1 = s[0].split("/")[0];
            String ip2 = s[1].split("/")[0];

            ipl.add(ip1);
            iph.add(ip2);
            loc.add(location);
            
        }
        Input.nextLine();
        for (int j=0;j<q;j++){
            String ipt = Input.nextLine();
            //System.out.println("inside second for");
            int flag = 0;

            for(int i=0;i<t;i++){
            long ipLo = ipToLong(InetAddress.getByName(ipl.get(i)));
            long ipHi = ipToLong(InetAddress.getByName(iph.get(i)));
            long ipToTest = ipToLong(InetAddress.getByName(ipt));

            if((ipToTest >= ipLo && ipToTest <= ipHi)){
                System.out.println(loc.get(i));
                flag=1;
                break;
                }   
            }
            if(flag==0){
                System.out.println("None");
            }
        }
    }
}